File: frontend/src/lib/contracts.ts
import { z } from "zod";

export const stageStatusSchema = z.enum(["pending", "running", "complete", "failed"]);
export const jobStatusSchema = z.enum(["queued", "downloading", "processing", "completed", "failed"]);

export const stageRecordSchema = z.object({
  key: z.string(),
  label: z.string(),
  status: stageStatusSchema,
  error: z.string().nullable(),
  updatedAt: z.string(),
});

export const jobRecordSchema = z.object({
  id: z.string(),
  sourceUrl: z.string().url(),
  status: jobStatusSchema,
  activeStage: z.string().nullable(),
  createdAt: z.string(),
  updatedAt: z.string(),
  stages: z.array(stageRecordSchema),
});

export const jobDetailResponseSchema = z.object({
  job: jobRecordSchema,
});

export type StageRecord = z.infer<typeof stageRecordSchema>;
export type JobRecord = z.infer<typeof jobRecordSchema>;
export type JobDetailResponse = z.infer<typeof jobDetailResponseSchema>;

File: backend/app/schemas/contracts.py
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, HttpUrl

StageStatus = Literal["pending", "running", "complete", "failed"]
JobStatus = Literal["queued", "downloading", "processing", "completed", "failed"]


class StageRecord(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    key: str
    label: str
    status: StageStatus
    error: Optional[str]
    updated_at: str = Field(alias="updatedAt")


class JobRecord(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    source_url: HttpUrl = Field(alias="sourceUrl")
    status: JobStatus
    active_stage: Optional[str] = Field(alias="activeStage")
    created_at: str = Field(alias="createdAt")
    updated_at: str = Field(alias="updatedAt")
    stages: list[StageRecord]


class JobDetailResponse(BaseModel):
    job: JobRecord
